_posts/2020-10-13-from-aligned-to-unaligned-checkpoints-part-1.md

```diff
@@ -0,0 +1,117 @@
+
+Despite all these great properties, Flink's checkpointing method has an Achilles Heel: the speed of a completed checkpoint is determined by the speed at which data flows through the application. When the application backpressures, the processing of checkpoints is backpressured as well (Appendix 1 recaps what is backpressure and why it can be a good thing). In such cases, checkpoints may take longer to complete or even time out completely.
+
+In Flink 1.11, the community introduced a first version of a new feature called "[unaligned checkpoints](https://ci.apache.org/projects/flink/flink-docs-stable/ops/state/checkpoints.html#unaligned-checkpoints)" that aims at solving this issue, while Flink 1.12 plans to further expand its functionality. In this two-series blog post, we discuss how Flink’s checkpointing mechanism has been modified to support unaligned checkpoints, how unaligned checkpoints work, and how this new mode impacts Flink users. In the first of the two posts, we start with a recap of the original checkpointing process in Flink, its core properties and issues under backpressure.
+
+
+## State in Streaming Applications
+
+Simply put, State is the information that you need to remember across events. Even the most trivial streaming applications are typically stateful because of their need to “remember” the exact position they are processing data from, in the form of a Kafka Partition Offset or a File Offset.
+In addition, many applications hold state internally as a way to support their internal operations, such as windows, aggregations, joins, or state machines.
+
+For the remainder of this article, we use the following example showing a streaming application consisting of four operators, each one holding some state
+
+<span>
+	<center>
+	<img vspace="8" style="width:75%" src="{{site.baseurl}}/img/blog/2020-10-13-from-aligned-to-unaligned-checkpoints-part-1/from-aligned-to-unaligned-checkpoints-part-1-imge-1.png" />
+	</center>
+</span>
+
+## State Persistence through Checkpoints
+
+Streaming applications are long-lived. They inevitably experience hardware and software failures but should, ideally, look from the outside as if no failure ever happened. Since applications are long-lived — and can potentially accumulate very large state —, recomputing partial results after failures can take quite some time and thus a way to persist and recover the (potentially very large) application state is necessary.
+
+Flink relies on its state checkpointing and recovery mechanism to implement such behavior, as shown in the figure below. Periodic checkpoints store a snapshot of the application’s state on some Checkpoint Storage (commonly an Object Store or Distributed File System, like S3, HDFS, GCS, Azure Blob Storage, etc.). When a failure is detected, the affected parts of the application are reset to the state of the latest checkpoint (either by a local reset or by loading the state from the checkpoint storage).
```

Comment:
```suggestion
Flink relies on its **state checkpointing and recovery mechanism** to implement such behavior, as shown in the figure below. Periodic checkpoints store a snapshot of the application’s state on some Checkpoint Storage (commonly an Object Store or Distributed File System, like S3, HDFS, GCS, Azure Blob Storage, etc.). When a failure is detected, the affected parts of the application are reset to the state of the latest checkpoint (either by a local reset or by loading the state from the checkpoint storage).
```

Reply:
Too much text with no styling, introducing some **bold** here and there to ease that.